tests/unit/package_managers/test_pip.py

```diff
@@ -3318,3 +3320,112 @@ def test_replace_external_requirements(
         assert replaced_file is not None
         assert replaced_file.template == expect_replaced
         assert replaced_file.abspath == requirements_path
+
+
+@pytest.mark.parametrize(
+    "packages, n_pip_packages",
+    [
+        ([], 0),
+        ([{"type": "gomod"}], 0),
+        ([{"type": "pip", "requirements_files": ["requirements.txt"]}], 1),
+        (
+            [
+                {"type": "pip", "requirements_files": ["requirements.txt"]},
+                {"type": "pip", "path": "foo", "requirements_build_files": []},
+            ],
+            2,
+        ),
+    ],
+)
+@mock.patch("cachi2.core.package_managers.pip._replace_external_requirements")
+@mock.patch("cachi2.core.package_managers.pip._resolve_pip")
+def test_fetch_pip_source(
+    mock_resolve_pip: mock.Mock,
+    mock_replace_requirements: mock.Mock,
+    packages: list[dict[str, Any]],
+    n_pip_packages: int,
+    tmp_path: Path,
+):
+    source_dir = tmp_path / "source"
+    output_dir = tmp_path / "output"
+    source_dir.mkdir()
+    source_dir.joinpath("foo").mkdir()
+
+    request = Request(source_dir=source_dir, output_dir=output_dir, packages=packages)
+
+    resolved_a = {
+        "package": {"name": "foo", "version": "1.0", "type": "pip"},
+        "dependencies": [
+            {"name": "bar", "version": "https://x.org/bar.zip", "type": "pip", "dev": False},
+            {"name": "baz", "version": "0.0.5", "type": "pip", "dev": True},
+        ],
+        "requirements": ["/package_a/requirements.txt", "/package_a/requirements-build.txt"],
+    }
+    resolved_b = {
+        "package": {"name": "spam", "version": "2.1", "type": "pip"},
+        "dependencies": [
+            {"name": "ham", "version": "3.2", "type": "pip", "dev": False},
+            {"name": "eggs", "version": "https://x.org/eggs.zip", "type": "pip", "dev": False},
+        ],
+        "requirements": ["/package_b/requirements.txt"],
+    }
+
+    replaced_file_a = ProjectFile(
+        abspath="/package_a/requirements.txt",
+        template="bar @ file://${output_dir}/deps/pip/...",
+    )
+    replaced_file_b = ProjectFile(
+        abspath="/package_b/requirements.txt",
+        template="eggs @ file://${output_dir}/deps/pip/...",
+    )
+
+    mock_resolve_pip.side_effect = [resolved_a, resolved_b]
+    mock_replace_requirements.side_effect = [replaced_file_a, None, replaced_file_b]
+
+    output = pip.fetch_pip_source(request)
+
+    expect_package_a = {
+        "name": "foo",
+        "version": "1.0",
+        "type": "pip",
+        "path": Path("."),
+        "dependencies": [
+            # TODO: add 'dev' attribute for non-gomod dependencies
+            {"name": "bar", "version": "https://x.org/bar.zip", "type": "pip"},  # "dev": False},
```

Comment:
I think I'll fix the `dev` key in a follow-up PR. There are at least two ways to do it and I don't want to make this PR too much more complex than it already is